P-UDN: support multiple cluster-subnets for Layer3 topology

l8huang opened the pull request:
## 📑 Description

Changes:
- Modify UDN/CUDN CRD schemas to allow adding multiple cluster-subents of
  the same IP family in Layer3 topology
- Reconcile network controllers to add new subnets to NodeAllocator.


Fixes #5377

## Additional Information for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [X] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [X] My code requires tests
- [X] if so, I have added and/or updated the tests as required
- [ ] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Layer3 user-defined networks support many subnets (limit raised, per-IP-family multi-subnet) with dynamic subnet additions and automatic per-node allocation and reconciliation.

* **Validation**
  * Per-field immutability added and stricter subnet rules enforced (no overlaps/duplicates, hostSubnet constraints, removal restrictions).

* **Reliability**
  * Nodes are requeued/reconciled when subnets expand so allocations propagate promptly.

* **Tests**
  * Extensive unit and e2e coverage for NAD lifecycle, allocation, updates, multi-subnet and validation scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
P-UDN: Support multiple cluster-subnets per IP family in Layer3 topology

- Modify UDN/CUDN CRD schemas to allow adding multiple cluster-subents of
  the same IP family in Layer3 topology
- Reconcile network controllers to add new subnets to NodeAllocator.

Fixes #5377

Signed-off-by: Lei Huang <leih@nvidia.com>

l8huang commented:
/retest-failed

l8huang commented:
/retest-failed

npinaeva commented on go-controller/pkg/crd/userdefinednetwork/v1/shared.go:
```diff
@@ -43,18 +46,18 @@ type Layer3Config struct {
-	// +kubebuilder:validation:MaxItems=2
+	// +kubebuilder:validation:MaxItems=400
 	// +required
-	// +kubebuilder:validation:XValidation:rule="size(self) != 2 || !isCIDR(self[0].cidr) || !isCIDR(self[1].cidr) || cidr(self[0].cidr).ip().family() != cidr(self[1].cidr).ip().family()", message="When 2 CIDRs are set, they must be from different IP families"
```
should we leave this rule for secondary networks?

l8huang replied:
I will try to preserve it for secondary networks.

l8huang replied:
this rule is added by for secondary networks.

Commit:
Add e2e test case for layer P-UDN multi-subnet support

This change adds test cases for:
- Pods can perform east/west traffic between nodes on different CIDR
- add subnet not affecting existing node subnet assignment
- add bad subnet should not cause change on existing NAD

Signed-off-by: Lei Huang <leih@nvidia.com>

l8huang commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -939,7 +955,11 @@ func (nInfo *userDefinedNetInfo) canReconcile(other NetInfo) bool {
+		if nInfo.topology != types.Layer3Topology {
+			return false
+		}
 	}
```
the HostSubnetLength check is done by CRD validation rule.

l8huang commented:
/retest-failed

Commit:
Update CRD validation rules and add corresponding e2e test cases

Signed-off-by: Lei Huang <leih@nvidia.com>

l8huang commented:
> I think we need to first merge #5397 to continue with the implementation?

please let me know if #5397 need further changes, thanks  CC @trozet

npinaeva commented on go-controller/pkg/clustermanager/clustermanager_suite_test.go:
```diff
@@ -1,13 +1,19 @@
+	// don't provide watchlist bookmarks and informer sync may time out.
+	if err := os.Setenv("KUBE_FEATURE_WatchListClient", "false"); err != nil {
+		t.Fatalf("Failed to set KUBE_FEATURE_WatchListClient: %v", err)
+	}
```
good point, i am wondering if we should put this to PrepareTestConfig to apply to all tests or some other place? we have quite some copies of this, including
```
if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(map[string]bool{"WatchListClient": false}); err != nil {
		t.Fatalf("Failed to disable WatchListClient feature gate: %v", err)
	}
```

l8huang replied:
Not sure how other test suites might be affected by moving this to PrepareTestConfig. I recommend removing the duplication in a dedicated PR.

npinaeva commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -479,8 +479,50 @@ func (ncc *networkClusterController) Cleanup() error {
 func (ncc *networkClusterController) Reconcile(netInfo util.NetInfo) error {
 	reconcilePendingPods := !ncc.ReconcilableNetInfo.EqualNADs(netInfo.GetNADs()...)
+
+	if ncc.nodeAllocator != nil {
+		oldSubnets := ncc.GetNetInfo().Subnets()
+		newSubnets := netInfo.Subnets()
+
+		// Find subnets that are in newSubnets but not in oldSubnets
+		addedSubnets := getNewSubnets(oldSubnets, newSubnets)
+		if len(addedSubnets) > 0 {
+			if err := ncc.nodeAllocator.AddSubnets(addedSubnets); err != nil {
+				klog.Errorf("Failed to add new subnets to node allocator for network %s: %v", ncc.GetNetworkName(), err)
+			}
+
+			// if some nodes failed to allocate a subnet, they most likely
+			// dropped out of the retry loop by the time a new subnet is added. Trigger a
+			// full reconcile for all nodes to ensure they get updated.
+			if err := objretry.RequeueAllNodes(ncc.watchFactory, netInfo, ncc.retryNodes); err != nil {
+				klog.Errorf("Failed to requeue all nodes for network %s: %v", ncc.GetNetworkName(), err)
+			}
+		}
+	}
```
I think that makes sense?

npinaeva replied:
@l8huang I think this needs to be fixed

l8huang replied:
updated

npinaeva commented on go-controller/pkg/retry/obj_retry.go:
```diff
@@ -860,3 +860,22 @@ func RequeuePendingPods(wf *factory.WatchFactory, netInfo util.NetInfo, retryPod
 }
+
+// RequeuePendingNodes renqueues all Pending nodes and triggers a retry immediately
+func RequeueAllNodes(wf *factory.WatchFactory, netInfo util.NetInfo, retryNodes *RetryFramework) error {
```
unused?

l8huang replied:
will remove it

l8huang replied:
removed

npinaeva commented on go-controller/pkg/testing/util.go:
```diff
@@ -188,3 +189,20 @@ func AddVTEPApplyReactor(fakeClient *vtepfake.Clientset) {
 	})
 }
+
+func BuildNAD(name, namespace string, network *ovncnitypes.NetConf) (*nadapi.NetworkAttachmentDefinition, error) {
```
I think that is very close to `GenerateNADWithConfig`?

l8huang replied:
yes, will try replace it with GenerateNAD() for GenerateNADWithConfig()

l8huang replied:
updated BuildNAD to reuse GenerateNADWithConfig

npinaeva commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1018,6 +1025,15 @@ func (nInfo *userDefinedNetInfo) TransitSubnets() []*net.IPNet {
+	if nInfo.topology == types.Layer3Topology {
+		nInfo.subnets = other.Subnets()
+	}
+}
```
won't it be handled by ` (l *mutableNetInfo) copyFrom(r *mutableNetInfo)`?

l8huang replied:
`(l *mutableNetInfo) copyFrom(r *mutableNetInfo)` only handles the embedded `mutableNetInfo`.

`nInfo.subnets = other.Subnets()` is required for updating outer userDefinedNetInfo

npinaeva replied:
oh I missed that subnets is now in both objects, i think it should be fully moved to the `mutableNetInfo` instead, or does it cause some issues I don't see?

l8huang replied:
not sure why is it

npinaeva replied:
so shouldn't subnets have been moved from `userDefinedNetInfo` to `mutableNetInfo`?

l8huang replied:
moved

npinaeva reviewed (commented):
This is not the full change though right? We have discussed in the OKEP all the other controllers that need to be updated

l8huang reviewed (commented):
This isn’t the full change. I’ll work on updating the other controllers and submit separate PRs where I can --  I’m not familiar with all of them, so might need help from other maintainers.

Thanks for reviewing this PR, I will update it based on your comments asap.

l8huang commented on test/e2e/network_segmentation.go:
```diff
@@ -2076,6 +2420,23 @@ func createManifest(namespace, manifest string) (func(), error) {
 	return cleanup, nil
 }
 
+func applyManifest(namespace, manifest string) (func(), error) {
+	path := "test-" + randString(5) + ".yaml"
+	if err := os.WriteFile(path, []byte(manifest), 0644); err != nil {
+		framework.Failf("Unable to write yaml to disk: %v", err)
+	}
+	cleanup := func() {
+		if err := os.Remove(path); err != nil {
+			framework.Logf("Unable to remove yaml from disk: %v", err)
+		}
+	}
+	_, err := e2ekubectl.RunKubectl(namespace, "apply", "-f", path)
+	if err != nil {
+		return cleanup, err
+	}
+	return cleanup, nil
+}
```
randString is defined in test/e2e/unidling.go

l8huang commented on go-controller/pkg/retry/obj_retry.go:
```diff
@@ -801,3 +801,22 @@ func RequeuePendingPods(wf *factory.WatchFactory, netInfo util.NetInfo, retryPod
+
+	retryNodes.RequestRetryObjs()
+	return nil
+}
```
fixed

l8huang committed:
networkconnect: handle P-UDN multi-subnet CNC router policies

Update ClusterNetworkConnect routing policy generation to create a
single logical router policy per destination network and IP family.
When a connected network has multiple same-family pod subnets, fold
them into one OR match instead of creating multiple policies with the
same DB identity.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

l8huang committed:
egressfirewall: update ACL when P-UDN adds overlapping subnet

Add unit coverage for the case where a primary UDN gains a new subnet
that overlaps an existing EgressFirewall CIDRSelector.

The test verifies that the initial ACL does not exclude the new subnet,
then updates the primary network subnets and confirms the reconciled ACL
adds the pod-subnet exclusion so east/west traffic is not affected by the
EgressFirewall rule.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

l8huang committed:
docs: clarify RouteAdvertisement handling for Layer3 multi-subnet UDN

Update the OKEP to note that no RouteAdvertisement code change is
required for Layer3 topology. FRR pod-subnet advertisements use node
subnet annotations, not the NAD cluster subnet list directly, and node
subnet annotation updates already trigger RA reconciliation.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

l8huang committed:
BGP: cover primary UDN multi-subnet route filtering for route importing

Add unit coverage for overlay mode route import when a primary UDN has
multiple cluster subnets. Verify that a BGP route contained in an
additional UDN subnet is ignored and not imported into OVN.

This confirms the existing route import logic accounts for all network
subnets when filtering pod-network routes in overlay mode.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

l8huang committed:
UDN: reconcile BGP isolation on primary UDN subnet changes

Queue advertised local nodes for reconciliation when a network subnet list
changes. This ensures BGP Network Isolation ACLs are rebuilt when a primary
Layer3 UDN gains an additional subnet.

Add unit coverage for both the reconcile trigger and the advertised network
isolation pass ACL update with the added subnet.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

npinaeva commented on go-controller/pkg/ovn/controller/networkconnect/topology.go:
```diff
@@ -986,36 +986,61 @@ func (c *Controller) ensureRoutingPoliciesOps(ops []ovsdb.Operation, cncName str
+	policyByFamily := map[string]*policyConfig{
+		"v4": policies[0],
+		"v6": policies[1],
+	}
```
I think this can be simplified, e.g. only have one container
```
policyByFamily := map[string]*policyConfig{
		"v4": {ipVersion: "ip4", ipFamily: "v4"},
		"v6": {ipVersion: "ip6", ipFamily: "v6"},
	}
```
and then iterate over it? You could then also remove `ipFamily` from the struct itself and get it from the map key

l8huang replied:
updated

npinaeva commented on go-controller/pkg/ovn/routeimport/route_import_test.go:
```diff
@@ -295,6 +320,27 @@ func Test_controller_syncNetwork(t *testing.T) {
 			},
 		},
+		{
+			name: "ignores primary UDN pod subnet routes from additional subnet in overlay mode",
```
should we add a no-overlay test too?

l8huang replied:
route import only ignores pod subnet routes in overlay mode, the new test case is for covering:

<route_import.go:49>
```
	if info.Transport() != types.NetworkTransportNoOverlay {
		// if the network is overlay mode, skip routes to the pod network
		ignoreSubnets = make([]*net.IPNet, len(info.Subnets()))
		for i, subnet := range info.Subnets() {
			ignoreSubnets[i] = subnet.CIDR
		}
	}
```

Existing no-overlay coverage already verifies pod subnet routes are imported -- test case "adds CUDN pod subnet routes in no-overlay mode"

npinaeva replied:
right, i think all tests in the context go in pairs, like
"ignores host subnet routes as necessary in overlay mode"
"adds host subnet routes as necessary in no-overlay mode"
and so on
So to continue this pattern, we could have 
"ignores primary UDN pod subnet routes from additional subnet in overlay mode"
"adds primary UDN pod subnet routes from additional subnet in no-overlay mode"?

l8huang replied:
added the no-overlay case

npinaeva reviewed (commented):
Thanks for adding more controllers to this PR!
I think another very important part is to start using multi-subnet L3 UDN as much as possible in our e2es, similar to what we did here https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5759/changes
So could you also go over the existing UDN e2e tests and try to update the subnets they are using to have multi-cidr config for primary L3?

l8huang committed:
egressip: reconcile routes when P-UDN subnets change

Trigger local node reconciliation when Egress IP is enabled and a primary
Layer3 UDN adds or removes cluster subnets.

Add unit coverage for the subnet-change reconcile path and verify Egress IP
route/no-reroute policy generation covers every primary Layer3 UDN subnet.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

l8huang commented:
/retest-failed

l8huang committed:
e2e: extend primary L3 UDN test cases with multiple subnets

Update primary Layer3 UDN/CUDN e2e coverage to use multi-subnet
configuration where the test is not specifically validating single-subnet
behavior. This broadens runtime coverage for existing network segmentation,
service, egress firewall, egress IP, endpoint slice mirroring, and route
advertisement flows.

Signed-off-by: Lei Huang <huang.f.lei@gmail.com>

npinaeva commented on test/e2e/e2e.go:
```diff
@@ -1665,20 +1665,30 @@ var _ = ginkgo.Describe("e2e ingress traffic validation", func() {
+				}
+				providerCtx.AddCleanUpFn(cleanupNodeIP)
+
+				// Cleanup any leftover address from a previous failed run before adding it again.
```
Does it mean the cleanup didn't work in the original test that added IPs?

l8huang replied:
I observed this issue when repeatedly running the test in my kind setup. Occasionally, the manually added IP was left behind, likely because the test was terminated before cleanup, and that blocked the next run.

CI always create new kind cluster, so it should not be a issue, these code are removed  to avoid confusion.

npinaeva commented on go-controller/pkg/ovn/base_network_controller.go:
```diff
@@ -201,11 +201,16 @@ func (oc *BaseNetworkController) reconcile(netInfo util.NetInfo, setNodeFailed f
+	reconcileEgressIP := config.OVNKubernetesFeature.EnableEgressIP &&
+		util.IsNetworkSegmentationSupportEnabled() &&
+		oc.IsPrimaryNetwork() &&
+		oc.TopologyType() == types.Layer3Topology
```
can't the subnet only change when all those condition are true anyway?
So would it be equivalent to just do
`reconcileSubnetChange := subnetsChanged && (isAdvertised || config.OVNKubernetesFeature.EnableEgressIP)`

l8huang replied:
yes, CRD validation only allows the subnet change for Primary Layer 3 UDN, here the check can be simplified. 

code updated.

npinaeva commented on test/e2e/cluster_network_connect.go:
```diff
@@ -329,8 +354,8 @@ func deleteCUDN(cudnName string) {
-// Pass empty strings for v4Subnet/v6Subnet to use defaults.
-func createPrimaryUDNWithSubnets(cs clientset.Interface, namespace, udnName, topology, v4Subnet, v6Subnet string) {
+// Pass an empty subnet list to use defaults.
+func createPrimaryUDNWithSubnets(cs clientset.Interface, namespace, udnName, topology string, subnets ...string) {
```
so it looks like it is only used by the egress firewall test with the new format, so you mind also updating CNC own network generation ? e.g. you could update `newTestNetworkSubnetsAllocator` to take multi-subnet as a bool argument and generate `172.{30+i}.0.0/23/24,172.{30+i+1}.0.0/16/24` subnet in the case?

l8huang replied:
CNC e2e updated.

npinaeva reviewed (commented):
nice work with the e2es, some final nits

l8huang reviewed (commented):
@npinaeva PR updated, PTAL, thanks

npinaeva reviewed (dismissed):
great work! unfortunately one last rebase is needed
